File: src/app/api/language.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  AsrModel,
  AlignModel,
  SourceLanguageOption,
} from "@/generated/bindings";

export async function listSourceLanguages(
  asrModel: AsrModel,
  alignModel: AlignModel,
): Promise<SourceLanguageOption[]> {
  return invoke("list_source_languages", { asrModel, alignModel });
}

File: src-tauri/src/commands/language.rs
use serde::Serialize;
use ts_rs::TS;

use crate::domain::language::LanguageTag;
use crate::domain::language_registry::LanguageRegistry;
use crate::services::preferences_types::{AlignModel, AsrModel};

#[derive(Serialize, TS)]
#[ts(export)]
pub struct SourceLanguageOption {
    pub tag: LanguageTag,
    pub label: String,
    pub short: String,
}

#[tauri::command]
pub fn list_source_languages(
    asr_model: AsrModel,
    align_model: AlignModel,
) -> Result<Vec<SourceLanguageOption>, String> {
    Ok(LanguageRegistry::supported_for(asr_model, align_model)
        .into_iter()
        .map(|m| SourceLanguageOption {
            tag: m.tag,
            label: m.display_name.to_string(),
            short: m.short_badge.to_string(),
        })
        .collect())
}
